src/index.ml

```diff
@@ -633,8 +633,15 @@ struct
                   Tbl.replace log.mem key value;
                   append_key_value log.io key value)
                 log_async.mem;
-              IO.flush log.io;
+
+              (* IO.flush can fail due to an exception in [auto_flush_callback]. *)
+              (try IO.flush log.io
+               with exn ->
+                 Mutex.unlock t.merge_lock;
+                 raise exn);
```

Comment:
That should be handled by the `Mutex.with_lock` you're wrapped in, right?

Reply:
True for the `rename_lock`, but not for the `merge_lock`. The `merge_lock` isn't properly scoped since it's taken by the main thread and dropped by the asynchronous one.

Reply:
Ah yes right, was too quick.